fem/qinterp/eval_hdiv.hpp

```diff
@@ -0,0 +1,462 @@
+                  {
+                     QQQ(qx,qy,qz,vd) = u[qz];
+                  }
+                  else if (Q_LAYOUT == QVectorLayout::byNODES)
+                  {
+                     y(qx,qy,qz,vd,e) = u[qz];
+                  }
+                  else // Q_LAYOUT == QVectorLayout::byVDIM
+                  {
+                     y(vd,qx,qy,qz,e) = u[qz];
+                  }
+               }
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+      if (FLAGS & (QuadratureInterpolator::PHYSICAL_VALUES |
+                   QuadratureInterpolator::PHYSICAL_MAGNITUDES))
+      {
+         MFEM_FOREACH_THREAD(qz,z,Q1D)
+         {
+            MFEM_FOREACH_THREAD(qy,y,Q1D)
+            {
+               MFEM_FOREACH_THREAD(qx,x,Q1D)
+               {
+                  real_t u_ref[DIM], u_phys[DIM];
+                  real_t J_loc[DIM*DIM];
+                  // Piola transformation: u_phys = J/det(J) * u_ref
+                  MFEM_UNROLL(DIM)
+                  for (int d = 0; d < DIM; d++)
+                  {
+                     u_ref[d] = QQQ(qx,qy,qz,d);
+                     MFEM_UNROLL(DIM)
+                     for (int sd = 0; sd < DIM; sd++)
+                     {
+                        J_loc[sd+DIM*d] = J(qx,qy,qz,sd,d,e);
+                     }
+                  }
+                  const real_t detJ = kernels::Det<DIM>(J_loc);
+                  kernels::Mult(DIM, DIM, J_loc, u_ref, u_phys);
+                  kernels::Set(DIM, 1, 1_r/detJ, u_phys, u_phys);
+                  if (FLAGS & QuadratureInterpolator::PHYSICAL_VALUES)
+                  {
+                     MFEM_UNROLL(DIM)
+                     for (int sd = 0; sd < DIM; sd++)
+                     {
+                        if (Q_LAYOUT == QVectorLayout::byNODES)
+                        {
+                           y(qx,qy,qz,sd,e) = u_phys[sd];
+                        }
+                        else // Q_LAYOUT == QVectorLayout::byVDIM
+                        {
+                           y(sd,qx,qy,qz,e) = u_phys[sd];
+                        }
+                     }
+                  }
+                  else if (FLAGS & QuadratureInterpolator::PHYSICAL_MAGNITUDES)
+                  {
+                     y(qx,qy,qz,0,e) = kernels::Norml2(DIM, u_phys);
+                  }
+               }
+            }
+         }
+         MFEM_SYNC_THREAD;
+      }
+   });
+}
+
+} // namespace quadrature_interpolator
+} // namespace internal
+
+/// @cond Suppress_Doxygen_warnings
+
+template<int DIM, QVectorLayout Q_LAYOUT, unsigned FLAGS, int D1D, int Q1D>
+QuadratureInterpolator::TensorEvalHDivKernelType
+QuadratureInterpolator::TensorEvalHDivKernels::Kernel()
+{
+   using namespace internal::quadrature_interpolator;
+   static_assert(DIM == 2 || DIM == 3, "only DIM=2 and DIM=3 are implemented!");
+   if (DIM == 2) { return EvalHDiv2D<Q_LAYOUT, FLAGS, D1D, Q1D>; }
+   else if (DIM == 3) { return EvalHDiv3D<Q_LAYOUT, FLAGS, D1D, Q1D>; }
```

Comment:
It looks like some compilers still need something like `else { MFEM_ABORT(""); }` to prevent 'missing return statement' warnings (despite the `static_assert` above).

Reply:
Should be addressed in #4721. 🤞